pkg/kn/commands/plugin/plugin_list.go

```diff
@@ -0,0 +1,190 @@
+// Copyright © 2019 The Knative Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plugin
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/knative/client/pkg/kn/commands"
+	"github.com/spf13/cobra"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+)
+
+// ValidPluginFilenamePrefixes controls the prefix for all kn plugins
+var ValidPluginFilenamePrefixes = []string{"kn"}
+
+// NewPluginListCommand creates a new `kn plugin list` command
+func NewPluginListCommand(p *commands.KnParams) *cobra.Command {
+	pluginFlags := PluginFlags{
+		IOStreams: genericclioptions.IOStreams{
+			In:     os.Stdin,
+			Out:    os.Stdout,
+			ErrOut: os.Stderr,
+		},
+	}
+
+	pluginListCommand := &cobra.Command{
+		Use:   "list",
+		Short: "List all visible plugin executables",
+		Long: `List all visible plugin executables.
+
+Available plugin files are those that are:
+- executable
+- begin with "kn-
+- anywhere on the path specfied in Kn's config pluginDir variable, which:
+  * defaults to $PATH if not specified
+  * can be overridden with the --plugin-dir flag`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			err := pluginFlags.complete(cmd)
+			if err != nil {
+				return err
+			}
+
+			err = pluginFlags.run()
+			if err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	pluginFlags.AddPluginFlags(pluginListCommand)
+
+	return pluginListCommand
+}
+
+// Private
+
+func (o *PluginFlags) complete(cmd *cobra.Command) error {
+	o.Verifier = &CommandOverrideVerifier{
+		Root:        cmd.Root(),
+		SeenPlugins: make(map[string]string, 0),
+	}
+
+	pluginPath := commands.PluginDir
+	if pluginPath == "$PATH" {
+		pluginPath = os.Getenv("PATH")
+	}
+
+	o.PluginPaths = filepath.SplitList(pluginPath)
+
+	return nil
+}
+
+func (o *PluginFlags) run() error {
+	pluginsFound := false
+	isFirstFile := true
+	pluginErrors := []error{}
+	pluginWarnings := 0
+
+	for _, dir := range uniquePathsList(o.PluginPaths) {
+		if dir == "" {
+			continue
+		}
+		files, err := ioutil.ReadDir(dir)
+		if err != nil {
+			if _, ok := err.(*os.PathError); ok {
+				fmt.Fprintf(o.ErrOut, "Unable read directory %q from your pluginDir: %v. Skipping...", dir, err)
```

Comment:
You could also argue that you’ve specified a directory for plugin that is not readable and important to tell you.

Reply:
I haven't specified anything. 

It's just that my PATH variable contains invalid directories. No shell complains about this when looking for executables, and I realized that just because kn told me.

So I have to fix path and my shell startup script for `kn` only to get rid of the warning. Sometimes this is not even possible for standard startup script which I can't write, so I have to live with a broken path with I can't fix. 

How would I get rid of this warning in such a situation ?

Reply:
I raised this point too.  Giving a warning about something that I might not have any control over will just annoy people. Besides, it's not an error (or even warning) condition to be in this state.

Reply:
Change your $PATH value.

Reply:
Why would should `kn` force it if no shell does when evaluating the path ? Are we really that dogmatic ?

Reply:
OK see what I have now is good enough.